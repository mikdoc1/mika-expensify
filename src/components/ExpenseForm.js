import React, {Component} from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import numeral from 'numeral';



class ExpenseForm extends Component {
    constructor (props) {
        super(props);
        this.state = {
            description: this.props.expense ? this.props.expense.description : '',
            amount: this.props.expense ? numeral(this.props.expense.amount / 100).format('0,0.00') : '',
            note: this.props.expense ? this.props.expense.note : '',
            isCalendarFocused: false,
            createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
            error: ''
        }
    }

    onAmountChange = e => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(({ amount }));
        }
    };
            
    onDescriptionChange = e => {
        const description = e.target.value;
        this.setState(({ description }));
    };

    onNoteChange = e => {
        const note = e.target.value;
        this.setState(({ note }));
    };

    onDateChange = createdAt => {
        if (createdAt) {
            this.setState(({ createdAt }))
        }    
    };

    onFocusChange = ({ focused }) => {
        this.setState(({ isCalendarFocused: focused}))
    };

    onSubmit = e => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(({ error: 'Please provide descripition and amount'}))
        } else {
            this.setState(({ error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };

    render() {
        return (
            <div>
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.onSubmit}>
                <input  type="text"
                        value={this.state.description}
                        placeholder="Desription"
                        onChange={this.onDescriptionChange}/>
                <input  type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}/>
                <SingleDatePicker   date={this.state.createdAt}
                                    onDateChange={this.onDateChange}
                                    focused={this.state.isCalendarFocused}
                                    onFocusChange={this.onFocusChange}
                                    numberOfMonths={1}
                                    isOutsideRange={() => false}/>
                <textarea   value={this.state.note}
                            onChange={this.onNoteChange}
                            placeholder="Add the note for your expenses (optional)">
                </textarea>
                <button>Add Expense</button>
            </form>
        </div>
        )   
    }
};

export default ExpenseForm
