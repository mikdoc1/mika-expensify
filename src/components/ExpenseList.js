import { connect } from 'react-redux';
import React from 'react';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => {
    let list = (
        <ul>
            {props.expenses.map(expense => {
                return <ExpenseListItem key={expense.description} {...expense}/>
            })}
        </ul>
    )
    if(props.expenses.length === 0) {
        list = <p>No expenses</p>
    }
    
    return (
        <div>
            {list}
        </div>    
    )
};

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};


export default connect(mapStateToProps)(ExpenseList)