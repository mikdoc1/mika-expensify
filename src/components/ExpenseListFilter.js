import React from 'react';
import { connect } from 'react-redux';
import * as actions from '../actions/filters';
import { DateRangePicker } from 'react-dates';
import { v4 as uuidv4 } from 'uuid';



class ExpenseListFilter extends React.Component {
    constructor (props) {
        super(props);
        this.state = {
            isCalendarFocused: null
        }
    };

    selectHandler = e => {
        const filterType = e.target.value
        if(filterType === 'amount') {
            this.props.onFilterByAmount()
        } else if (filterType === 'date') {
            this.props.onFilterByDate()
        }
    }

    render = () => {

        return (
            <div>
            <input  type="text" 
                    value={this.props.filters.text}
                    onChange={(e) => this.props.onFilterByText(e.target.value)}/>
            <select onChange={this.selectHandler} value={this.props.filters.sortBy}>
                <option value='date'>Date</option>
                <option value='amount'>Amount</option>
            </select>
            <DateRangePicker    startDate={this.props.filters.startDate}
                                startDateId={uuidv4()}
                                endDate={this.props.filters.startDate}
                                endDateId={uuidv4()}
                                numberOfMonths={1}
                                isOutsideRange={() => false}
                                showClearDates={true}
                                onDatesChange={({ startDate, endDate }) => { 
                                                                
                                                                // this.props.onDatesStartChange(startDate); 
                                                                // this.props.onDatesEndChange(endDate)
                                                             }}
                                focusedInput={this.state.isCalendarFocused}
                                onFocusChange={(calFocused) => this.setState({ isCalendarFocused: calFocused})}/>
                                                                
                                
        </div>
        )
    }
};

const mapStateToProps = state => {
    return {
        filters: state.filters
    }  
};


const mapDispatchToProps = dispatch => {
    return {
      onFilterByText: (text) => dispatch(actions.setTextFilter(text)),
      onFilterByAmount: () => dispatch(actions.sortByAmount()),
      onFilterByDate: () => dispatch(actions.sortByDate()),
    //   onDatesStartChange: (startDate) => dispatch(actions.setStartDate(startDate)),
    //   onDatesEndChange: (endDate) => dispatch(actions.setEndDate(endDate))
    }
}
export { ExpenseListFilter }

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter)