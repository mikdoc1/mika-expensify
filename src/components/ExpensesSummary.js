import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import expensesTotal from '../selectors/expenses-total';
import selectExpesnes from '../selectors/expenses';

export const ExpensesSummary = ({expensesCount, expensesTotal}) => {
    
    const expenseWord = expensesCount === 0 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');

    return ( 
        <div>
            <h2>Viewing {expensesCount} {expenseWord} tottaling {formattedExpensesTotal}</h2>
        </div>
    )
};

const mapStateToProps = state => {
    const visibleExpenses = selectExpesnes(state.expenses, state.filters);

    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: expensesTotal(visibleExpenses)
    }
};


export default connect(mapStateToProps)(ExpensesSummary);