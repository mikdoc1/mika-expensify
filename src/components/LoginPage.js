import React from 'react';
import { connect } from 'react-redux';
import { loginStart } from '../actions/auth';

export const LoginPage = (props) => {
    return (
        <div>
            <button onClick={props.onLogin}>Login</button>
        </div>
    )
};

const mapDispatchToProps = dispatch => {
    return {
        onLogin: () => dispatch(loginStart())
    }
}


export default connect(null, mapDispatchToProps)(LoginPage)