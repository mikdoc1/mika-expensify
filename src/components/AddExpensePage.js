import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import * as actions from '../actions/expenses'

export const AddExpensePage = (props) => {

  return (
    <div>
    <h1>Add Expense</h1>
    <ExpenseForm onSubmit={(expense) => {
                            props.onStartAddExpense(expense); 
                            props.history.push('/')
                          }}/>            
  </div>
  )
};



const matDispatchToProps = dispatch => {
  return {
    onStartAddExpense: (expense) => dispatch(actions.startAddExpense(expense))
  }
}

export default connect(null, matDispatchToProps)(AddExpensePage);
