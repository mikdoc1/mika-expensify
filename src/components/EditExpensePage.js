import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import * as actions from '../actions/expenses';

export const EditExpensePage = (props) => {

  return (
    <div>
      <ExpenseForm  onSubmit={(expense) => {
                        
                        
                        props.onEditExpense(expense, props.expense.id);
                        props.history.push('/')
                    }}
                    expense={props.expense}/>
      <button onClick={() => {
  
                        props.onDeleteExpense(props.expense.id); 
                        props.history.push('/')
                      }}>Remove</button>
    </div>
  );
};

const mapStateToProps = (state, props) => {

  return {
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
  }
}
const mapDispatchToProps = dispatch => {
  return {
    onEditExpense: (expense, id) => dispatch(actions.startEditExpense(id, expense)),
    onDeleteExpense: (id) => dispatch(actions.startRemoveExpense({id}))
  }
}


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);


