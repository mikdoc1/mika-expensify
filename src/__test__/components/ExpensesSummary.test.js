import { ExpensesSummary }  from '../../components/ExpensesSummary';
import React from 'react';
import { shallow } from 'enzyme';

test('should correctly render ExpensesSummary with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={200}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary with multy expenses', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={5} expensesTotal={1200.34}/>);
    expect(wrapper).toMatchSnapshot();
});