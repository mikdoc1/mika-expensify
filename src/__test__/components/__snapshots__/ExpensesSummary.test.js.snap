// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <h2>
    1
     
    expenses
     tottaling 
    $2.00
  </h2>
</div>
`;

exports[`should correctly render ExpensesSummary with multy expenses 1`] = `
<div>
  <h2>
    5
     
    expenses
     tottaling 
    $12.00
  </h2>
</div>
`;
